Add AIAgent as subclass of SoftwareAgent

bact opened the pull request:
Draft of `/AI/AIAgent` which is subclass of `/Core/SoftwareAgent`.

<img width="601" height="481" alt="ai-aiagent" src="https://github.com/user-attachments/assets/d932fd2d-fe14-4533-9f10-be4b5acaea24" />

Apart from base properties from `/Core/SoftwareAgent`, it will have additional properties as proposed by @ElyasRashno, @stevenc-stb, @cngo20 and others:

- `/Core/suppliedBy` (from Core profile)
- https://github.com/spdx/spdx-3-model/issues/1093#issue-3342194071 -> `/Software/enabledByDefault`
- #1088 -> `/Software/inputModality`
- #1087 -> `/Software/outputModality`
  - Note that the `Modality` vocab is the same as in PR #1110 (copied here to make it passed the parser check)
- #1089 -> `/AI/agentCapability`
- #1084 -> `/AI/agentMemoryMode`
  - Few of the suggested values look a bit overlap with `/AI/promptPattern` here? https://github.com/spdx/spdx-3-model/pull/1110/files
- #1064 -> `/AI/automationLevel`

## Relationship types

Some properties are converted from the original proposal to existing/new relationship types to maximise benefits of existing ecosystem:

- "runs" (from https://lists.spdx.org/g/spdx-ai/message/338) -> `invokedBy`
  - Using `invokedBy` relationship, we can utilised the existing `delegatedTo` ("on behalf of") relationship too
  - `runsOn`, proposed from Hardware profile, can be repurposed / considered
- #1086 -> `usesTool`
- #1079 -> `hasPersistentMemory`
  - The new relationship type could potentially be used in other non-AIAgent context as well.
  - If we feel this one is better to be property, can do that as well - to be discussed
  - Alternatively, we can pick the name `usesLongTermMemoryStore` - to be discussed

```mermaid
graph LR;
  D(AIPackage) -->|invokedBy| B(AIAgent)
  B(AIAgent) -->|invokedBy| C(Person)
  B(AIAgent) -->|hasPersistentMemory| A(Package)
  B(AIAgent) -->|usesTool| E(Package)
```

(Please see a table of existing relationship types at
https://docs.google.com/spreadsheets/d/1g9TophkZ2tWhTPMGGHqGOUurzZvOdkBtBJCvyBsa8kc/edit?usp=sharing )

Note that more properties can be added later, this is just a skeleton.

bact committed:
Add AIAgent as subclass of SoftwareAgent

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add hasPersistentMemory

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add AIAgent diagram

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

stevenc-stb commented on model/AI/Classes/AIAgent.md:
```diff
@@ -0,0 +1,43 @@
+## Metadata
+
+- name: AIAgent
+- SubclassOf: /Core/Agent
```
should an AIAgent be a sub class of SoftwareAgent?

bact replied:
Oops. Updated. Thanks for noticing that. Fixed.

bact committed:
Agent -> SoftwareAgent

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: stevenc-stb <192034142+stevenc-stb@users.noreply.github.com>

bact committed:
Add enabledByDefault

Suggested by @cngo20 in https://github.com/spdx/spdx-3-model/issues/1093#issue-3342194071

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: cngo20 <63925596+cngo20@users.noreply.github.com>

bact committed:
Add enabledByDefault definition

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: cngo20 <63925596+cngo20@users.noreply.github.com>

bact committed:
Add inputModality + outputModality

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
Co-Authored-By: ElyasRashno <42498039+elyasrashno@users.noreply.github.com>

bact committed:
Rename diagram files

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

goneall commented:
@bact @rgopikrishnan91 - similar comment, please update on the status

bact committed:
Update Modality vocab

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Update gitignore

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Update with ISO verbal forms

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

ElyasRashno committed:
Added agent capabilities

ElyasRashno committed:
added agentExternalTools

ElyasRashno committed:
added agentMemoryMode

bact commented:
Btw, please look at definitions in files:
https://github.com/spdx/spdx-3-model/pull/1113/files

For AI Agent, it is currently:

> An AI agent is a software agent that uses artificial intelligence techniques
to guide its behavior. It typically incorporates a perception-action cycle,
utilizing models and heuristics to perform goal-directed activities.

bact committed:
Update model/AI/Classes/AIAgent.md

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix Range

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix Range

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Remove changes in gitignore

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix typo

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix typo mincount -> minCount

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix typo mincount -> minCount

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix typo mincount -> minCount

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Merge pull request #2 from ElyasRashno/ai-agent-elyas-extension

Ai agent elyas extension

bact committed:
Update formatting and singular form

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact commented:
Merged agentCapability, etc. by @ElyasRashno from PR https://github.com/bact/spdx-3-model/pull/2

bact committed:
Update diagram

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Merge branch 'spdx:develop' into add-ai-agent

bact committed:
Merge branch 'develop' into add-ai-agent

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

kestewart commented:
Please let us know if this should be 3.2 or not.

bact commented:
With the current state of direction within the AI WG, the fundamental design has changed.

We can close this and a new PR with the updated design can be opened later.
